app/models/taxonomy.rb

```diff
@@ -28,6 +28,10 @@ class Taxonomy < ApplicationRecord
   has_many :subnets, :through => :taxable_taxonomies, :source => :taxable, :source_type => 'Subnet'
   has_many :auth_sources, :through => :taxable_taxonomies, :source => :taxable, :source_type => 'AuthSource'
 
+  has_many :puppetclasses, :through => :environments
+  has_many :hosts, :through => :environments
```

Comment:
taxonomies already have many hosts directly, they don't need to be defined through environments

Reply:
It's doesn't seem to be able to find the hosts association properly though, without this association `Organization.find(ID).reports` fails:

```ruby
[24] pry(main)> Organization.first.reports
2018-12-05T09:46:06 [D|sql|]   Organization Load (0.7ms)  SELECT  "taxonomies".* FROM "taxonomies" WHERE "taxonomies"."type" IN ('Organization') ORDER BY "taxonomies"."title" ASC LIMIT $1  [["LIMIT", 1]]
ActiveRecord::HasManyThroughAssociationNotFoundError: Could not find the association :hosts in model Taxonomy
from /home/bastilian/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/activerecord-5.2.1/lib/active_record/reflection.rb:912:in `check_validity!'
```

Reply:
hmm, the association is defined on the child classes, so maybe the reports association should also be defined there?

Reply:
Yes, that works. Though, should we maybe think about moving the `has_many_hosts` one level up, or is it in the child classes for a specific reason?

Reply:
I think it's there so the reverse is calculated correctly (i.e. on host there is both organization_id and location_id, not taxonomy_id)